### PR TITLE
Alternative crosslink data construction, take 1

### PR DESCRIPTION
**Pull request by vbuterin**
Properties:

* Optimal data packing
* Fixed depth
* All data is encoded; makes fraud proofs not-too-difficult

Weaknesses:

* Uses SSZ serialization.
* Positions of blocks dynamic, so accesses require an extra Merkle branch

### CODE REVIEW

File: specs/core/1_shard-data-chains.md [hunk 1]
```diff
@@ -76,7 +79,7 @@ The following types are defined, mapping into `DomainType` (little endian):
 
 | Name | Value |
 | - | - |
-| `PLACEHOLDER` | `2**32` |
+| `PLACEHOLDER` | `2**3` |
```
Comment (hwwhww) [inline specs/core/1_shard-data-chains.md]: I changed it to `2**3` for testing.

File: specs/core/1_shard-data-chains.md [hunk 2]
```diff
@@ -107,7 +110,7 @@ class ShardBlock(Container):
     shard: Shard
     beacon_chain_root: Bytes32
     parent_root: Bytes32
-    data: ShardBlockBody
+    body: ShardBlockBody
```
Comment (hwwhww) [inline specs/core/1_shard-data-chains.md]: Also hotfix in #1298.

File: specs/core/1_shard-data-chains.md [hunk 3]
```diff
@@ -247,36 +250,37 @@ def verify_shard_attestation_signature(state: BeaconState,
...
+    return zpad(serialize(get_shard_header(block)), BYTES_PER_SHARD_BLOCK_HEADER) + serialize(block.body)
+```
+
+### `compute_crosslink_data`
+
+```python
+def compute_crosslink_data(blocks: Sequence[ShardBlock]) -> Bytes32:
+    """
+    Flattens a series of blocks. Designed to be equivalent to SSZ serializing a list of
+    flattened blocks with one empty block at the end as a termination marker.
```
Comment (hwwhww) [inline specs/core/1_shard-data-chains.md]: Could you elaborate what do you mean by "one empty block at the end" here?

File: specs/core/1_shard-data-chains.md [hunk 4]
```diff
@@ -247,36 +250,37 @@ def verify_shard_attestation_signature(state: BeaconState,
     )
 ```
 
+### `flatten_block`
+
+```python
+def flatten_block(block: ShardBlock) -> Bytes:
+    return zpad(serialize(get_shard_header(block)), BYTES_PER_SHARD_BLOCK_HEADER) + serialize(block.body)
```
Comment (hwwhww) [inline specs/core/1_shard-data-chains.md]: I added `serialize()` here, correct?

File: specs/core/1_shard-data-chains.md (hunk 3 above)
Comment (hwwhww) [reply specs/core/1_shard-data-chains.md]: Could you elaborate what's the main advantage of this one over just using SSZ `serialize` like:
```python
class CrosslinkingBlocks:
    data: List[ShardBlock, SHARD_SLOTS_PER_EPOCH * MAX_EPOCHS_PER_CROSSLINK]

assert len(blocks) <= SHARD_SLOTS_PER_EPOCH * MAX_EPOCHS_PER_CROSSLINK
crosslinking_blocks = CrosslinkingBlocks(data=*(blocks))
return serialize(crosslinking_blocks)
```
?

File: specs/core/1_shard-data-chains.md [hunk 5]
```diff
@@ -247,36 +250,37 @@ def verify_shard_attestation_signature(state: BeaconState,
...
+```python
+def compute_crosslink_data(blocks: Sequence[ShardBlock]) -> bytes:
+    """
+    Flattens a series of blocks. Designed to be equivalent to SSZ serializing a list of
+    flattened blocks with one empty block at the end as a termination marker.
+    """
+    positions = [4 * len(blocks)]
+    bodies = []
+    for block in blocks:
+        bodies.append(flatten_block(block))
```
Comment (djrtwo) [inline specs/core/1_shard-data-chains.md]: Doesn't flattening the block here reduce our ability to construct proofs about components of a block through the crosslink root? Still possible but not as clean due to flattening to bytes

**Comment by dankrad** [comment]
> Uses SSZ serialization.

I wouldn't see that as a weakness, if it is the right tool for the job, I think we should be able to bring it into consensus.

**Comment by djrtwo** [comment]
> I wouldn't see that as a weakness, if it is the right tool for the job, I think we should be able to bring it into consensus.

agreed.

**Comment by djrtwo** [comment]
@vbuterin This can be closed as it was superseded by #1294, right?

**Comment by djrtwo** [comment]
Closing -- superseded by #1294

**Pull request closed without merging**